### PR TITLE
Implement `stack bench --benchmark-arguments`

### PR DESCRIPTION
**Pull request by markus1189**
Implements #496.

I renamed the `boptsTestArgs` to `boptsAdditionalArgs` and use that field for both testing and benchmarking.  I also chose to use `--benchmark-arguments` instead of `--benchmark-options` because for tests it is `--test-arguments`.

With this `stack bench --benchmark-arguments='-o $benchmark.html --csv $benchmark.csv'` works as expected.

### CODE REVIEW

**Comment by snoyberg** [comment]
Looks good at first blush. I'm on my phone right now and probably won't be able to properly review until Sunday. If someone else wants to review and merge I have no objection.

File: src/Stack/Build/Execute.hs [hunk 1]
```diff
@@ -883,8 +883,9 @@ singleBench ac ee task =
             config <- asks getConfig
             cabal (console && configHideTHLoading config) ["build"]
 
+        let args = boptsAdditionalArgs (eeBuildOpts ee)
         announce "benchmarks"
-        cabal False ["bench"]
+        cabal False ["bench","--benchmark-options=" <> (intercalate " " args) <> ""]
```
Comment (snoyberg) [inline src/Stack/Build/Execute.hs]: This won't properly handle arguments that have spaces in them, we'd need some kind of quoting mechanism to be used instead. I'm also unsure what the `<> ""` at the end is trying to do.

Comment (markus1189) [inline src/Stack/Build/Execute.hs]: The `<> ""` was a check for your watchful eyes, obviously :)

**Comment by markus1189** [comment]
You're right about the need for a quoting mechanism, any suggestions?

**Comment by snoyberg** [comment]
I can't think of a good prebuilt solution. @chrisdone are you aware of anything?

**Comment by manny-fp** [comment]
What about using multiple `--benchmark-option` (singular) arguments instead?

**Comment by manny-fp** [comment]
Alternatively, could just take `--benchmark-arguments` as a string rather than parsing it into a list, and pass it verbatim to `--benchmark-options`?  But I think I like the former suggestion better, that way the quoting behaviour is consistent with others in stack and not platform-dependent.

**Comment by markus1189** [comment]
Actually things like
```
$ stack bench --benchmark-arguments='--csv $benchmark.csv -o $benchmark.html -v 2'
```
work, we get `args = ["--csv","$benchmark.csv","-o","$benchmark.html","-v","2"]` which (curiously to me) works with `cabal False ["bench","--benchmark-options=" <> unwords args]`

The problems occur if one of the arguments has spaces, as noted correctly above by @snoyberg:

```
$ stack bench --benchmark-arguments='--csv "my $benchmark".csv -o "my $benchmark.html" -v 2'
```
The `args` list looks like this: `["--csv","my $benchmark.csv","-o","my $benchmark.html","-v","2"]`.

I don't think we can use `--benchmark-option`, because we still have the problem about quoting the `my $benchmark.html` filename containing spaces...

**Comment by markus1189** [comment]
Okay I just switched out the `argsOption` parser for optparse's `strOption`, as `argsOption` unescapes the string.  Now it works:

```
$ stack bench --benchmark-arguments='--csv "my $benchmark.csv" -o "my $benchmark.html" -v 2'
```
results in `args = ["--csv \"my $benchmark.csv\" -o \"my $benchmark.html\" -v 2"]`.

I chose to still rename the field in `BuildOpts` to `boptsAdditionalArgs` and use it for benchmarks, in theory we could also keep the old `boptsTestArgs` and add a special field `boptsBenchArgs`.

What speaks for the first is that we keep the number of fields in `BuildOpts` minimal, what speaks for the second is that `boptsTestArgs` is `[String]` which does not make sense for benchmarks because we use only one string and pass it on, though we can just use a singleton list as I am doing right now.

TL;DR: Rename and reuse `boptsTestArgs` or add special field `boptsBenchArgs` to `BuildOpts`?

**Comment by snoyberg** [comment]
I'll throw a new option into the mix: add fields to the `FinalAction` type so that test and benchmark specific options do not appear in `BuildOpts`. I've already done that for rerunning tests, I think it would make sense to do that more generally. This could look like:

```haskell
data FinalAction
    = DoTests TestOpts
    | DoBenchmarks BenchmarkOpts
    | DoNothing
```

Anyone see a downside to that?

**Comment by chrisdone** [comment]
I thought about adding stuff to `FinalAction` a couple times, and extending the args parser so that each action can have its own parser. Not opposed.

**Comment by snoyberg** [comment]
@markus1189 Not intending to hoist extra work onto you that you didn't intend to sign up for, but are you interested in taking the stab at making this change? It should be mostly mechanical and pretty well isolated from other things.

**Comment by markus1189** [comment]
No problem at all, but thanks for asking anyway.

Summary:
- add `TestOpts` to `DoTests` and `BenchmarkOpts` to `DoBenchmarks`
- remove the previously introduced `boptsAdditionalArgs` in favor of the fields in `FinalAction`

and I also took the liberty to move the flags for `--no-run-tests` and `coverage` to `TestOpts` as well as fix #537 because I was touching it anyway.

Do you want me to squash the commits into one with a summary in the body?

File: src/Stack/Build/Types.hs [hunk 2]
```diff
@@ -320,19 +315,28 @@ defaultBuildOpts = BuildOpts
...
     , boptsOnlySnapshot = False
-    , boptsCoverage = False
     , boptsFileWatch = False
     , boptsKeepGoing = Nothing
-    , boptsNoTests = False
     }
 
+-- | Options for the 'FinalAction' 'DoTests'
+data TestOpts =
+  TestOpts {toRerunTests :: !Bool -- | Whether successful tests will be run gain
```
Comment (snoyberg) [inline src/Stack/Build/Types.hs]: I think you want to use `-- ^` instead

**Comment by snoyberg** [comment]
I just had a minor comment about haddock syntax (`-- ^` instead of `-- |`). Other than that, this looks great, thank you!

**Comment by markus1189** [comment]
Good catch. Changed the haddock for `TestOpts` and `BenchmarkOpts`.
